lib/Frontend/Frontend.cpp

```diff
@@ -190,107 +192,111 @@ bool CompilerInstance::setUpInputs(Optional<unsigned> codeCompletionBufferID) {
+       llvm::sys::path::filename(input.file()) == "main.swift")) {
+    assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
+    MainBufferID = *bufferID;
+  }
 
-    if (isPrimary) {
-      assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
-      PrimaryBufferID = bufferID;
-    }
+  if (input.isPrimary()) {
+    assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
+    PrimaryBufferID = *bufferID;
   }
+  return false;
 }
 
-bool CompilerInstance::setUpForFile(StringRef fileName, bool isPrimary) {
-  if (fileName.empty())
-    return false;
-  // FIXME: Working with filenames is fragile, maybe use the real path
-  // or have some kind of FileManager.
-  using namespace llvm::sys::path;
-  if (Optional<unsigned> existingBufferID =
-          SourceMgr.getIDForBufferIdentifier(fileName)) {
-    if (isInSILMode() ||
-        (isInputSwift() && filename(fileName) == "main.swift")) {
-      assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
-      MainBufferID = existingBufferID.getValue();
-    }
-
-    if (isPrimary) {
-      assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
-      PrimaryBufferID = existingBufferID.getValue();
+Optional<unsigned> CompilerInstance::getRecordedBufferID(const InputFile &input,
+                                                         bool &failed) {
+  if (!input.buffer()) {
+    if (Optional<unsigned> existingBufferID =
+            SourceMgr.getIDForBufferIdentifier(input.file())) {
+      return existingBufferID;
     }
-
-    return false; // replaced by a memory buffer.
   }
+  std::pair<std::unique_ptr<llvm::MemoryBuffer>,
+            std::unique_ptr<llvm::MemoryBuffer>>
+      buffers = getInputBufferAndModuleDocBufferIfPresent(input);
 
-  // Open the input file.
-  using FileOrError = llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>;
-  FileOrError inputFileOrErr = llvm::MemoryBuffer::getFileOrSTDIN(fileName);
-  if (!inputFileOrErr) {
-    Diagnostics.diagnose(SourceLoc(), diag::error_open_input_file, fileName,
-                         inputFileOrErr.getError().message());
-    return true;
+  if (!buffers.first.get()) {
+    failed = true;
+    return None;
   }
 
-  if (serialization::isSerializedAST(inputFileOrErr.get()->getBuffer())) {
-    llvm::SmallString<128> ModuleDocFilePath(fileName);
-    llvm::sys::path::replace_extension(ModuleDocFilePath,
-                                       SERIALIZED_MODULE_DOC_EXTENSION);
-    FileOrError moduleDocOrErr =
-        llvm::MemoryBuffer::getFileOrSTDIN(ModuleDocFilePath.str());
-    if (!moduleDocOrErr &&
-        moduleDocOrErr.getError() != std::errc::no_such_file_or_directory) {
-      Diagnostics.diagnose(SourceLoc(), diag::error_open_input_file, fileName,
-                           moduleDocOrErr.getError().message());
-      return true;
-    }
+  if (serialization::isSerializedAST(buffers.first.get()->getBuffer())) {
```

Comment:
Hm, seems unfortunate to do this test twice. Maybe trying to do everything with a `std::pair` is too clever?

(Also, `std::unique_ptr` has an `operator->`, so no need for the `get()` here either.)

Reply:
I understand about repeating the test, but, short of creating a new class to represent the outcome of the processing and then collecting all the side-effects into one place--a further reorg I like but don't want to do now--I don't see a clean way to eliminate it. It would be good to revisit this code later. I'll add in a FIXME.